src/devicecontroller/DefaultDeviceController.ts

```diff
@@ -293,131 +210,65 @@ export default class DefaultDeviceController
-    }
-
-    // Only Chrome needs this fix.
-    if (recreateAudioContext && !this.browserBehavior.requiresContextRecreationForAudioWorklet()) {
-      this.logger.debug('Browser does not require audio context recreation hack.');
-      recreateAudioContext = false;
+      return undefined;
     }
-
-    // Only need to do this if either device has an audio worklet.
-    if (recreateAudioContext && !this.transform && isAudioTransformDevice(device)) {
-      this.logger.debug('Neither device is a transform. No need to recreate audio context.');
-      recreateAudioContext = false;
+    if (this.getAudioInputInProgress) {
+      throw new Error('Another start audio input is in progress');
     }
+    this.getAudioInputInProgress = true;
 
-    if (recreateAudioContext) {
-      this.logger.info('Recreating audio context when selecting new device.');
-
-      /* istanbul ignore else */
-      if (this.transform) {
-        /* istanbul ignore else */
-        if (this.transform.nodes) {
-          this.transform.nodes.end.disconnect();
-          this.transform.nodes = undefined;
-        }
-        this.transform = undefined;
-      }
-
-      /* istanbul ignore else */
-      if (this.audioInputSourceNode) {
-        this.audioInputSourceNode.disconnect();
-        this.audioInputSourceNode = undefined;
-      }
+    try {
+      if (isAudioTransformDevice(device)) {
+        // N.B., do not JSON.stringify here — for some kinds of devices this
+        // will cause a cyclic object reference error.
+        this.logger.info(`Choosing transform input device ${device}`);
 
-      /* istanbul ignore else */
-      if (this.audioInputDestinationNode) {
-        this.audioInputDestinationNode.disconnect();
-        this.audioInputDestinationNode = undefined;
+        await this.chooseAudioTransformInputDevice(device);
+      } else {
+        this.logger.info(`Choosing intrinsic audio input device ${device}`);
+        this.removeTransform();
+        await this.chooseInputIntrinsicDevice('audio', device);
       }
-
-      DefaultDeviceController.closeAudioContext();
+      this.trace('startAudioInputDevice', device, `success`);
+      return this.useWebAudio
+        ? this.getMediaStreamDestinationNode().stream
+        : this.activeDevices['audio'].stream;
+    } catch (error) {
+      throw error;
+    } finally {
+      this.getAudioInputInProgress = false;
     }
+  }
 
-    if (isAudioTransformDevice(device)) {
-      // N.B., do not JSON.stringify here — for some kinds of devices this
-      // will cause a cyclic object reference error.
-      this.logger.info(`Choosing transform input device ${device}`);
-
-      await this.chooseAudioTransformInputDevice(device);
-    } else {
-      this.logger.info(`Choosing intrinsic audio input device ${device}`);
-      this.removeTransform();
-      await this.chooseInputIntrinsicDevice('audio', device, false);
-      this.trace('chooseAudioInputDevice', device, `success`);
+  stopAudioInput(): void {
```

Comment:
How about making `stop()` async too to make start/stop symmetrical?

Reply:
Made them async.